SpatialGDK/Source/SpatialGDK/Private/EngineClasses/GridBasedLBStrategy.cpp

```diff
@@ -0,0 +1,101 @@
+// Fill out your copyright notice in the Description page of Project Settings.
+
+
+#include "GridBasedLBStrategy.h"
+
+UGridBasedLBStrategy::UGridBasedLBStrategy()
+	: Super()
+	, Rows(1)
+	, Cols(1)
+	, WorldWidth(10000.f)
+	, WorldHeight(10000.f)
+	, EdgeFuzziness(1.f)
+{
+}
+
+void UGridBasedLBStrategy::Init(const class USpatialNetDriver* InNetDriver)
+{
+	Super::Init(InNetDriver);
+
+	for (uint32 i = 1; i <= Rows * Cols; i++)
+	{
+		VirtualWorkerIds.Add(i);
+	}
+
+	// TODO(my) Copied this logic wholesale from prototype
+	const float WorldWidthMin = -(WorldWidth / 2.f);
+	const float WorldHeightMin = -(WorldHeight / 2.f);
+
+	const float ColumnWidth = WorldWidth / Cols;
+	const float RowHeight = WorldHeight / Rows;
+
+	for (uint32 Col = 0; Col < Cols; ++Col)
+	{
+		for (uint32 Row = 0; Row < Rows; ++Row)
+		{
+			FVector2D Min(WorldWidthMin + (Col * ColumnWidth), WorldHeightMin + (Row * RowHeight));
+			FVector2D Max(Min.X + ColumnWidth, Min.Y + RowHeight);
+			FBox2D Cell(Min, Max);
+
+			// Fuzzy cell edges to avoid floating point and boundary errors (2cm overlap).
+			// This implies that an actor will be in 2 or more cells when crossing
+			// boundaries. This is OK, since we first ask the current authoritative
+			// worker if it should relinquish authority, and it will not do that until
+			// the actor leaves its cell, including the fuzzy boundary. When we then ask
+			// which worker should get authority, a different worker will always be
+			// selected (assuming we have complete coverage over the world.)
+			Cell = Cell.ExpandBy(EdgeFuzziness);
+
+			WorkerCells.Add(MoveTemp(Cell));
+		}
+	}
+}
+
+TArray<uint32> UGridBasedLBStrategy::GetVirtualWorkerIds() const
+{
+	return VirtualWorkerIds;
+}
+
+bool UGridBasedLBStrategy::ShouldRelinquishAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		return false;
+	}
+
+	const FVector2D Actor2DLocation = FVector2D(Actor.GetActorLocation());
```

Comment:
I wonder if we should use spatial position instead of unreal location for this.

Reply:
I'd be in favour of using the unreal position unless we find that doesn't work for large enough worlds. If users are expected to write strategies then it would be ideal if they don't touch position.

Reply:
I was thinking of the Position component. I learned a couple of days ago that (I think) the player controller's unreal location isn't updated as the player moves. If we want to stick with the UE location, then we need to check for owned components and PCs, like we do when updating the Position component.